userspace/libsinsp/sinsp_ppm_sc.cpp

```diff
@@ -217,6 +217,95 @@ std::unordered_set<uint32_t> sinsp::enforce_sys_ppm_sc_set(std::unordered_set<ui
 	return ppm_sc_set;
 }
 
+void sinsp::enforce_optimized_ppm_sc_set(std::unordered_set<uint32_t>& ppm_sc_set, bool print)
+{
+	std::unordered_set<uint32_t> ppm_sc_set_filter_evttypes = ppm_sc_set;
+
+	/* Enforce syscalls solely needed for minimal sinsp state built-up and life-cycle management (Falco use case). */
+	std::unordered_set<uint32_t> base_thread_state_set = {
```

Comment:
Can't we start from `enforce_sinsp_state_ppm_sc`?
In that case, we can be sure that any syscall needed by sinsp enrichment is actually captured.

Reply:
Primary goal of this PR is to provide an alternative to `enforce_sinsp_state_ppm_sc` with the `--optimized-syscalls` option (opt-in, default to false) 🙃 .
Thinking behind this is that current approach seems to add too many unnecessary syscalls (because of too generic `UF_NEVER_DROP` or `EF_MODIFIES_STATE` flags).

Consider the following (attempting to brain dump my current understanding):

- For example, consider a scenario where Falco rules are only around `spawned_process` and `open*` syscalls -> hence no need to tap into any network related syscalls at all or other fds.
- Many `EF_MODIFIES_STATE` do `EF_CREATES_FD`, but consider a scenario where this syscall isn't part of a Falco rule, hence don't need to add those fds -> for the most part most fd related syscalls contain the needed fd info in the actual live event. Only for example for `connect`, `accept*` and `listen` syscalls absolutely need to also trace `socket` syscall etc to get the ip tuples. However, interested in learning about concrete examples where this assessment is false -> then extend new alternative `enforce_optimized_ppm_sc_set` option.
- Very carefully reviewed the parser and the current sets is what came out, but I planned more detailed analyses. The intended new minimum set should truly only address adding or updating fields that are used to create logs when another syscall triggers (e.g. opened a file, but also need current uid of the process). 

[The removing and destroying parts should be sufficiently addressed (it's only the proc exit sched events and close syscall)]

Reply:
So, i'd say:
* either we enforce the sinsp state ppm_sc
* or we DO NOT enforce them, and expect the consumer to specify additional syscalls to be tracked, aside from the ruleset ones

If you see, all APIs in that source file  but `enforce_simple_ppm_sc` do not enforce the sinsp state ppm sc!

Reply:
Of course, Falco will eventually enforce the `sinsp_state_ppm_sc` though :/

Reply:
I think that sooner or later we will need to expose a config option to let users choose the syscalls to be additionally tracked, like
`additional_syscalls: [ "open", "close", "execve" ]` and when the option is set, we should avoid enforcing the sinsp state ppm sc.

Reply:
@FedeDP summarized all new suggestions in Falco PR https://github.com/falcosecurity/falco/pull/2361.

- Re `additional_syscalls` option, I like it, gives more control and even more options to end users
- Re sinsp state enforcement, obviously since I opened this would say we really need this new approach (once finalized and proven to be 100% technically sound), but we can support both: (1) a more efficient sinsp state enforcement option (this one) and (2) the more conservative option (the existing API). Conservative one can be default and the new one opt-in via a flag.

Reply:
Note however that `UF_NEVER_DROP` are likely to be dropped from the enforced set (i think we will eventually drop both the `UF_NEVER_DROP` and `UF_ALWAYS_DROP` flags sooner or later).
In the end, our enforced state ppm sc will just be the list of `EF_MODIFIES_STATE` events, not that big!